File: BackEnd/JavaScript/expressshop_prismaorm/src/addresses/address.repository.js
import { prisma } from "../prisma.js";

function addressSelect() {
  return {
    id: true,
    userId: true,
    label: true,
    recipientName: true,
    phone: true,
    zipCode: true,
    address: true,
    addressDetail: true,
    isDefault: true,
    createdAt: true,
    updatedAt: true,
  };
}

export function resetDefaultAddresses(userId) {
  return prisma.address.updateMany({ where: { userId }, data: { isDefault: false } });
}

export function findAddresses(userId) {
  return prisma.address.findMany({
    where: { userId },
    select: addressSelect(),
    orderBy: [{ isDefault: "desc" }, { id: "asc" }],
  });
}

export function createAddress(data) {
  return prisma.address.create({ data, select: addressSelect() });
}

export function findAddress(userId, addressId) {
  return prisma.address.findFirst({
    where: { id: Number(addressId), userId },
  });
}

export function updateAddressById(addressId, data) {
  return prisma.address.update({
    where: { id: Number(addressId) },
    data,
    select: addressSelect(),
  });
}

export function deleteAddressById(userId, addressId) {
  return prisma.address.deleteMany({
    where: { id: Number(addressId), userId },
  });
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/address/AddressRepository.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.address

import java.time.Instant

data class AddressRecord(
    val id: Int,
    val userId: Int,
    val label: String,
    val recipientName: String,
    val phone: String,
    val zipCode: String,
    val address: String,
    val addressDetail: String?,
    val isDefault: Boolean,
    val createdAt: Instant,
)

data class NewAddress(
    val label: String,
    val recipientName: String,
    val phone: String,
    val zipCode: String,
    val address: String,
    val addressDetail: String?,
    val isDefault: Boolean,
)

data class AddressUpdate(
    val label: String? = null,
    val recipientName: String? = null,
    val phone: String? = null,
    val zipCode: String? = null,
    val address: String? = null,
    val addressDetail: String? = null,
    val isDefault: Boolean? = null,
)

interface AddressRepository {
    fun listAddresses(userId: Int): List<AddressRecord>

    fun findAddressById(
        userId: Int,
        addressId: Int,
    ): AddressRecord?

    fun createAddress(
        userId: Int,
        newAddress: NewAddress,
    ): AddressRecord

    fun updateAddress(
        userId: Int,
        addressId: Int,
        update: AddressUpdate,
    ): AddressRecord

    fun deleteAddress(
        userId: Int,
        addressId: Int,
    )
}
